mne/stats/cluster_level.py

```diff
@@ -1729,3 +1737,310 @@ def summarize_clusters_stc(
+    pivot_df["evoked"] = pivot_df[conditions[0]] - pivot_df[conditions[1]]
+
+    # Optional: Clean up the DataFrame
+    pivot_df = pivot_df[["subject_index", "channel", "timepoint", "evoked"]]
+
+    # check if formula is present
+    if formula is not None:
+        formulaic = _soft_import(
+            "formulaic", purpose="set up Design Matrix"
+        )  # soft import (not a dependency for MNE)
+
+        # for the paired t-test y is the difference between conditions
+        # X is the design matrix with a column with 1s and 0s for each participant
+        # Create the design matrix using formulaic
+        y, X = formulaic.model_matrix(formula, pivot_df)
+    else:
+        raise ValueError(
+            "Formula is required and needs to be a string in Wilkinson notation."
+        )
+
+    # now prep design matrix for input into MNE cluster function
+    # cluster functions expects channels as list dimension
+    y_for_cluster = y.values.reshape(-1, n_channels, n_timepoints).transpose(0, 2, 1)
+
+    adjacency, _ = find_ch_adjacency(df["evoked"][0].info, ch_type="eeg")
+
+    # define stat function and threshold
+    stat_fun, threshold = _check_fun(
+        X=y_for_cluster, stat_fun=None, threshold=None, tail=0, kind="within"
+    )
+
+    # Run the cluster-based permutation test
+    T_obs, clusters, cluster_p_values, H0 = _permutation_cluster_test(
+        [y_for_cluster],
+        n_permutations=n_permutations,
+        threshold=threshold,
+        stat_fun=stat_fun,
+        tail=tail,
+        n_jobs=n_jobs,
+        adjacency=adjacency,
+        max_step=max_step,  # maximum distance between samples (time points)
+        exclude=exclude,  # exclude no time points or channels
+        step_down_p=step_down_p,  # step down in jumps test
+        t_power=t_power,  # weigh each location by its stats score
+        out_type=out_type,
+        check_disjoint=check_disjoint,
+        buffer_size=buffer_size,  # block size for chunking the data
+        seed=seed,
+    )
+
+    print(f"smallest cluster p-value: {min(cluster_p_values)}")
+
+    return ClusterResult(T_obs, clusters, cluster_p_values, H0)
+
+
+def unpack_time_and_channels(df: pd.DataFrame = None) -> pd.DataFrame:
+    """
+    Extract timepoints and channels and convert to long.
+
+    Parameters
+    ----------
+    df : pd.DataFrame
+        DataFrame in wide format.
+
+    Returns
+    -------
+    df_long : pd.DataFrame
+        DataFrame in long format.
+    """
+    # Extracting all necessary data using list comprehensions for better performance
+    long_format_data = [
+        {
+            "condition": row["condition"],
+            "subject_index": row["subject_index"],
+            "channel": channel,
+            "timepoint": timepoint,
+            "value": row["data"][channel, timepoint],
+        }
+        for idx, row in df.iterrows()
+        for channel in range(row["data"].shape[0])
+        for timepoint in range(row["data"].shape[1])
+    ]
```

Comment:
there may be a (likely much faster) built-in way to do this in Pandas. maybe https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.explode.html ?

Reply:
Is this what you're looking for perhaps?

https://pandas.pydata.org/docs/reference/api/pandas.wide_to_long.html

But I didn't spend a long time looking, ignore my comment if it's not helpful

Reply:
I checked both pandas functions and tried the but it didn't seem that they are useful in this specific case. But I am happy to be convinced otherwise.